# 10-K Financial Data Analysis: Microsoft, Tesla & Apple

## Objective
Analyze key financial data extracted from SEC EDGAR 10-K filings for **Microsoft**, **Tesla**, and **Apple** over the last three fiscal years (FY2022–FY2024). This analysis will identify trends and insights that could inform the development of an AI-powered financial chatbot.

## Methodology
1. **Data Extraction** — Financial figures manually extracted from 10-K filings on SEC EDGAR
2. **Data Loading** — CSV loaded into pandas DataFrame
3. **Trend Analysis** — Year-over-year growth rates calculated for all metrics
4. **Comparative Analysis** — Cross-company comparisons and insights
5. **Conclusions** — Key findings for AI chatbot development

### Data Sources
- [Microsoft 10-K (SEC EDGAR)](https://www.sec.gov/cgi-bin/browse-edgar?action=getcompany&CIK=0000789019&type=10-K)
- [Tesla 10-K (SEC EDGAR)](https://www.sec.gov/cgi-bin/browse-edgar?action=getcompany&CIK=0001318605&type=10-K)
- [Apple 10-K (SEC EDGAR)](https://www.sec.gov/cgi-bin/browse-edgar?action=getcompany&CIK=0000320193&type=10-K)

---
## Step 1: Import Libraries and Load Data

In [1]:
import pandas as pd
import warnings
warnings.filterwarnings('ignore')

# Load extracted financial data
df = pd.read_csv('financial_data.csv')
print(f'Dataset Shape: {df.shape}')
print(f'\nFinancial Data (All figures in USD):')
df

Dataset Shape: (9, 7)

Financial Data (All figures in USD):


,Company,Fiscal Year,Total Revenue,Net Income,Total Assets,Total Liabilities,Cash Flow from Operating Activities
0,Microsoft,2022,198270000000,72738000000,364840000000,198298000000,89035000000
1,Microsoft,2023,211915000000,72361000000,411976000000,205753000000,87582000000
2,Microsoft,2024,245122000000,88136000000,512163000000,243686000000,118548000000
3,Tesla,2022,81462000000,12556000000,82338000000,36440000000,14724000000
4,Tesla,2023,96773000000,14974000000,106618000000,43009000000,13256000000
5,Tesla,2024,97690000000,7091000000,122070000000,48261000000,11527000000
6,Apple,2022,394328000000,99803000000,352755000000,302083000000,122151000000
7,Apple,2023,383285000000,96995000000,352583000000,290437000000,110543000000
8,Apple,2024,391035000000,93736000000,364980000000,308030000000,118254000000


---
## Step 2: Data Formatting — Convert to Billions for Readability

In [2]:
# Create a readable version in billions
df_display = df.copy()
numeric_cols = ['Total Revenue', 'Net Income', 'Total Assets', 'Total Liabilities', 'Cash Flow from Operating Activities']

df_display['Revenue ($B)'] = (df['Total Revenue'] / 1e9).round(2)
df_display['Net Income ($B)'] = (df['Net Income'] / 1e9).round(2)
df_display['Total Assets ($B)'] = (df['Total Assets'] / 1e9).round(2)
df_display['Total Liabilities ($B)'] = (df['Total Liabilities'] / 1e9).round(2)
df_display['Operating Cash Flow ($B)'] = (df['Cash Flow from Operating Activities'] / 1e9).round(2)

display_cols = ['Company', 'Fiscal Year', 'Revenue ($B)', 'Net Income ($B)', 'Total Assets ($B)', 'Total Liabilities ($B)', 'Operating Cash Flow ($B)']
print('\nFinancial Data (in Billions USD):')
print(df_display[display_cols].to_string())


Financial Data (in Billions USD):
     Company  Fiscal Year  Revenue ($B)  Net Income ($B)  Total Assets ($B)  Total Liabilities ($B)  Operating Cash Flow ($B)
0  Microsoft         2022        198.27            72.74             364.84                  198.30                     89.04
1  Microsoft         2023        211.92            72.36             411.98                  205.75                     87.58
2  Microsoft         2024        245.12            88.14             512.16                  243.69                    118.55
3      Tesla         2022         81.46            12.56              82.34                   36.44                     14.72
4      Tesla         2023         96.77            14.97             106.62                   43.01                     13.26
5      Tesla         2024         97.69             7.09             122.07                   48.26                     11.53
6      Apple         2022        394.33            99.80             352.76        

---
## Step 3: Year-over-Year Growth Analysis

In [3]:
# Calculate Year-over-Year growth rates
df['Revenue Growth (%)'] = df.groupby('Company')['Total Revenue'].pct_change() * 100
df['Net Income Growth (%)'] = df.groupby('Company')['Net Income'].pct_change() * 100
df['Asset Growth (%)'] = df.groupby('Company')['Total Assets'].pct_change() * 100
df['Liabilities Growth (%)'] = df.groupby('Company')['Total Liabilities'].pct_change() * 100
df['Operating CF Growth (%)'] = df.groupby('Company')['Cash Flow from Operating Activities'].pct_change() * 100

growth_cols = ['Company', 'Fiscal Year', 'Revenue Growth (%)', 'Net Income Growth (%)', 'Asset Growth (%)', 'Liabilities Growth (%)', 'Operating CF Growth (%)']
print('\n=== Year-over-Year Growth Rates (%) ===')
print(df[growth_cols].dropna().round(2).to_string())


=== Year-over-Year Growth Rates (%) ===
     Company  Fiscal Year  Revenue Growth (%)  Net Income Growth (%)  Asset Growth (%)  Liabilities Growth (%)  Operating CF Growth (%)
1  Microsoft         2023                6.88                  -0.52             12.92                    3.76                    -1.63
2  Microsoft         2024               15.67                  21.80             24.32                   18.44                    35.36
4      Tesla         2023               18.80                  19.26             29.49                   18.03                    -9.97
5      Tesla         2024                0.95                 -52.64             14.49                   12.21                   -13.04
7      Apple         2023               -2.80                  -2.81             -0.05                   -3.86                    -9.50
8      Apple         2024                2.02                  -3.36              3.52                    6.06                     6.98


---
## Step 4: Profitability and Efficiency Ratios

In [4]:
# Calculate key financial ratios
df['Profit Margin (%)'] = (df['Net Income'] / df['Total Revenue'] * 100).round(2)
df['Debt-to-Asset Ratio'] = (df['Total Liabilities'] / df['Total Assets']).round(2)
df['OCF-to-Revenue (%)'] = (df['Cash Flow from Operating Activities'] / df['Total Revenue'] * 100).round(2)

ratio_cols = ['Company', 'Fiscal Year', 'Profit Margin (%)', 'Debt-to-Asset Ratio', 'OCF-to-Revenue (%)']
print('\n=== Key Financial Ratios ===')
print(df[ratio_cols].to_string())


=== Key Financial Ratios ===
     Company  Fiscal Year  Profit Margin (%)  Debt-to-Asset Ratio  OCF-to-Revenue (%)
0  Microsoft         2022              36.69                 0.54               44.91
1  Microsoft         2023              34.15                 0.50               41.33
2  Microsoft         2024              35.96                 0.48               48.36
3      Tesla         2022              15.41                 0.44               18.07
4      Tesla         2023              15.47                 0.40               13.70
5      Tesla         2024               7.26                 0.40               11.80
6      Apple         2022              25.31                 0.86               30.98
7      Apple         2023              25.31                 0.82               28.84
8      Apple         2024              23.97                 0.84               30.24


---
## Step 5: Comparative Summary by Company

In [5]:
# Average metrics by company
summary = df.groupby('Company').agg({
    'Total Revenue': 'mean',
    'Net Income': 'mean',
    'Profit Margin (%)': 'mean',
    'Debt-to-Asset Ratio': 'mean',
    'Cash Flow from Operating Activities': 'mean'
}).round(2)

summary.columns = ['Revenue ($B)', 'Net Income ($B)', 'Profit Margin (%)', 'Debt-to-Asset', 'OCF ($B)']
summary['Revenue ($B)'] = (summary['Revenue ($B)'] / 1e9).round(2)
summary['Net Income ($B)'] = (summary['Net Income ($B)'] / 1e9).round(2)
summary['OCF ($B)'] = (summary['OCF ($B)'] / 1e9).round(2)

print('\n=== Average Metrics by Company (FY2022-FY2024) ===')
print(summary.to_string())

# Revenue ranking for latest year
latest = df[df['Fiscal Year'] == 2024][['Company', 'Total Revenue']].copy()
latest['FY2024 Revenue ($B)'] = (latest['Total Revenue'] / 1e9).round(2)
print('\n=== FY2024 Revenue Ranking ===')
print(latest[['Company', 'FY2024 Revenue ($B)']].sort_values('FY2024 Revenue ($B)', ascending=False).to_string())


=== Average Metrics by Company (FY2022-FY2024) ===


           Revenue ($B)  Net Income ($B)  Profit Margin (%)  Debt-to-Asset  OCF ($B)
Company                                                                             
Apple            389.55            96.84              24.86           0.84    116.98
Microsoft        218.44            77.74              35.60           0.51     98.39
Tesla             91.98            11.54              12.71           0.41     13.17

=== FY2024 Revenue Ranking ===
     Company  FY2024 Revenue ($B)
8      Apple               391.04
2  Microsoft               245.12
5      Tesla                97.69


---
## Step 6: Trend Analysis — Revenue Over 3 Years

In [6]:
# Revenue pivot
rev_pivot = df.pivot_table(index='Company', columns='Fiscal Year', values='Total Revenue')
rev_pivot = (rev_pivot / 1e9).round(2)
rev_pivot['3-Year Change (%)'] = ((rev_pivot[2024] - rev_pivot[2022]) / rev_pivot[2022] * 100).round(2)
print('\n=== Revenue Trends ($B) ===')
print(rev_pivot.to_string())

# Net Income pivot
ni_pivot = df.pivot_table(index='Company', columns='Fiscal Year', values='Net Income')
ni_pivot = (ni_pivot / 1e9).round(2)
ni_pivot['3-Year Change (%)'] = ((ni_pivot[2024] - ni_pivot[2022]) / ni_pivot[2022] * 100).round(2)
print('\n=== Net Income Trends ($B) ===')
print(ni_pivot.to_string())

# Cash Flow pivot
cf_pivot = df.pivot_table(index='Company', columns='Fiscal Year', values='Cash Flow from Operating Activities')
cf_pivot = (cf_pivot / 1e9).round(2)
cf_pivot['3-Year Change (%)'] = ((cf_pivot[2024] - cf_pivot[2022]) / cf_pivot[2022] * 100).round(2)
print('\n=== Operating Cash Flow Trends ($B) ===')
print(cf_pivot.to_string())


=== Revenue Trends ($B) ===
Fiscal Year    2022    2023    2024  3-Year Change (%)
Company                                               
Apple        394.33  383.28  391.04              -0.83
Microsoft    198.27  211.92  245.12              23.63
Tesla         81.46   96.77   97.69              19.92



=== Net Income Trends ($B) ===
Fiscal Year   2022   2023   2024  3-Year Change (%)
Company                                            
Apple        99.80  97.00  93.74              -6.07
Microsoft    72.74  72.36  88.14              21.17
Tesla        12.56  14.97   7.09             -43.55



=== Operating Cash Flow Trends ($B) ===
Fiscal Year    2022    2023    2024  3-Year Change (%)
Company                                               
Apple        122.15  110.54  118.25              -3.19
Microsoft     89.04   87.58  118.55              33.14
Tesla         14.72   13.26   11.53             -21.67


---
## Step 7: Key Findings and Insights

### Microsoft — Strong and Accelerating Growth
- **Revenue** grew 23.6% over 3 years, accelerating from 6.9% (FY23) to 15.7% (FY24)
- **Net Income** surged 21.8% in FY2024 after being flat in FY2023, driven by AI/cloud growth
- **Profit margin** remains the highest of all three companies at ~36%
- **Operating cash flow** jumped 35% in FY2024 — strongest cash generation growth
- **Insight**: Microsoft's AI investments (Copilot, Azure OpenAI) are clearly paying off

### Tesla — Revenue Growth Stalling, Profitability Declining
- **Revenue** growth collapsed from 18.8% (FY23) to just 0.95% (FY24)
- **Net Income** dropped 52.7% in FY2024 — the sharpest decline of all three companies
- **Profit margin** halved from 15.4% to 7.3% due to aggressive price cuts
- **Operating cash flow** declined for two consecutive years (-10% then -13%)
- **Insight**: Price war strategy is hurting margins significantly; growth model under pressure

### Apple — Stable Revenue, Gradual Margin Compression
- **Revenue** dipped 2.8% in FY2023 then recovered modestly (+2%) in FY2024
- **Net Income** has declined for two consecutive years (-2.8%, -3.4%)
- **Profit margin** stable around 24-25% but trending slightly downward
- **Debt-to-Asset ratio** is highest at 0.84 — Apple uses significant leverage
- **Insight**: Hardware saturation visible; Services segment is critical for future growth

---

## Implications for AI Financial Chatbot Development

Based on this analysis, an AI-powered financial chatbot should be designed to:

1. **Automate trend detection** — Identify accelerating/decelerating growth patterns automatically
2. **Calculate key ratios** — Profit margins, debt ratios, cash flow efficiency on demand
3. **Cross-company comparison** — Allow users to compare companies across multiple metrics
4. **Alert on anomalies** — Flag significant YoY changes (e.g., Tesla's 52.7% net income drop)
5. **Natural language summaries** — Convert raw financial data into human-readable insights
6. **Historical context** — Provide multi-year trend context for any single data point

### Key Data Points the Chatbot Must Handle:
- Revenue, Net Income, Total Assets, Total Liabilities, Operating Cash Flow
- Year-over-year growth rates
- Profit margins, debt ratios, cash flow efficiency
- Multi-year trend analysis and cross-company benchmarking

In [7]:
# Final summary table
final = pd.DataFrame({
    'Company': ['Microsoft', 'Tesla', 'Apple'],
    '3Y Rev Growth (%)': [23.63, 19.93, -0.83],
    '3Y NI Growth (%)': [21.17, -43.55, -6.07],
    'Avg Profit Margin (%)': [35.60, 12.71, 24.86],
    'Avg Debt/Asset': [0.51, 0.41, 0.84],
    'FY24 Revenue ($B)': [245.12, 97.69, 391.04]
})

print('\n' + '='*40)
print('  FINAL SUMMARY TABLE')
print('='*40)
print(final.to_string(index=False))
print('\nAnalysis Complete. Notebook ready for export.')


  FINAL SUMMARY TABLE
  Company  3Y Rev Growth (%)  3Y NI Growth (%)  Avg Profit Margin (%)  Avg Debt/Asset  FY24 Revenue ($B)
Microsoft              23.63             21.17                  35.60            0.51             245.12
    Tesla              19.93            -43.55                  12.71            0.41              97.69
    Apple              -0.83             -6.07                  24.86            0.84             391.04

Analysis Complete. Notebook ready for export.
